src/ext/coms_curl.c

```diff
@@ -0,0 +1,246 @@
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <sys/time.h>
+#include <time.h>
+
+#include "compatibility.h"
+#include "coms.h"
+#include "coms_curl.h"
+#include "env_config.h"
+#include "vendor_stdatomic.h"
+
+#define HOST_FORMAT_STR "http://%s:%u/v0.4/traces"
+#define DEFAULT_FLUSH_INTERVAL (5000)
+#define DEFAULT_FLUSH_AFTER_N_REQUESTS (10)
+#define DEFAULT_AGENT_CONNECT_TIMEOUT (100)
+#define DEFAULT_AGENT_TIMEOUT (500)
+
+struct _writer_loop_data_t {
+    pthread_t thread;
+    pthread_mutex_t mutex;
+    pthread_cond_t condition;
+    _Atomic(BOOL_T) shutdown;
+    _Atomic(BOOL_T) send;
+    _Atomic(uint32_t) request_counter;
+    _Atomic(uint32_t) requests_since_last_flush;
+};
+
+inline static uint32_t get_flush_interval() {
+    return ddtrace_get_uint32_config("DD_TRACE_AGENT_FLUSH_INTERVAL", DEFAULT_FLUSH_INTERVAL);
+}
+
+inline static uint32_t get_flush_after_n_requests(){
+    return ddtrace_get_uint32_config("DD_TRACE_AGENT_FLUSH_AFTER_N_REQUESTS", DEFAULT_FLUSH_AFTER_N_REQUESTS);
+}
+
+inline static void curl_set_timeout(CURL *curl) {
+    uint32_t agent_timeout = ddtrace_get_uint32_config("DD_TRACE_AGENT_TIMEOUT", DEFAULT_AGENT_TIMEOUT);
+    curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT_MS, agent_timeout);
+}
+
+inline static void curl_set_connect_timeout(CURL *curl) {
+    uint32_t agent_connect_timeout =
+        ddtrace_get_uint32_config("DD_TRACE_AGENT_CONNECT_TIMEOUT", DEFAULT_AGENT_CONNECT_TIMEOUT);
+    curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT_MS, agent_connect_timeout);
+}
+
+inline static void curl_set_hostname(CURL *curl) {
+    char *hostname = ddtrace_get_c_string_config("DD_AGENT_HOST");
+    int64_t port = ddtrace_get_int_config("DD_TRACE_AGENT_PORT", 8126);
+    if (port <= 0 || port > 65535) {
+        port = 8126;
+    }
+
+    if (hostname) {
+        size_t agent_url_len =
+            strlen(hostname) + sizeof(HOST_FORMAT_STR) + 10;  // port digit allocation + some headroom
+        char *agent_url = malloc(agent_url_len);
+        snprintf(agent_url, agent_url_len, HOST_FORMAT_STR, hostname, (uint32_t)port);
+
+        curl_easy_setopt(curl, CURLOPT_URL, agent_url);
+        ddtrace_env_free(hostname);
+        free(agent_url);
+    } else {
+        curl_easy_setopt(curl, CURLOPT_URL, "http://localhost:8126/v0.4/traces");
+    }
+}
+
+inline static struct timespec deadline_in_ms(uint32_t ms) {
+    struct timespec deadline;
+    struct timeval now;
+
+    gettimeofday(&now, NULL);
+    uint32_t sec = ms / 1000UL;
+    uint32_t msec = ms % 1000UL;
+    deadline.tv_sec = now.tv_sec + sec;
+    deadline.tv_nsec = ((now.tv_usec + 1000UL * msec) * 1000UL);
+
+    // carry over full seconds from nsec
+    deadline.tv_sec += deadline.tv_nsec / (1000 * 1000 * 1000);
+    deadline.tv_nsec %= (1000 * 1000 * 1000);
+
+    return deadline;
+}
+
+inline static BOOL_T curl_debug() {
+    return ddtrace_get_bool_config("DD_TRACE_DEBUG_CURL_OUTPUT", FALSE);
```

Comment:
there were 3 invocations of exact same `ddtrace_get_bool_config`. Refactored into a function since 3 is the magic number for me :)
Its not memoized though. If you mean the static keyword - its only meant to make this function invisible to other compilation units (or whatever is called in C lingo).

Reply:
I like your "magical rule of 3"! :) So if we call it one more time... MEMOIZE!!!!!1 Right? 😆

Reply:
Im not sure - on C level getenv should be quite fast. For tight loops it still could be beneficial to avoid executions  but otherwise the cost should be negligible. (famoust last words ;) )